synapse/storage/databases/state/bg_updates.py

```diff
@@ -182,69 +216,35 @@ def _get_state_groups_from_groups_txn(
+                # The `*_` rest syntax is to ignore the `state_group` column which we
+                # only select in the optimized case
+                typ, state_key, event_id, *_ = row
+                key = (intern_string(typ), intern_string(state_key))
+                # Deal with the potential duplicate (type, state_key) pairs from the
+                # SQLite specific query above. We only want to use the first row which
+                # is from the greatest state group (most-recent) because that is that
+                # applicable state in that state group.
+                if key not in results[group]:
                     results[group][key] = event_id
-        else:
-            max_entries_returned = state_filter.max_entries_returned()
-
-            where_clause, where_args = state_filter.make_sql_filter_clause()
-            # Unless the filter clause is empty, we're going to append it after an
-            # existing where clause
-            if where_clause:
-                where_clause = " AND (%s)" % (where_clause,)
-
-            # We don't use WITH RECURSIVE on sqlite3 as there are distributions
-            # that ship with an sqlite3 version that doesn't support it (e.g. wheezy)
-            for group in groups:
-                next_group: Optional[int] = group
-
-                while next_group:
-                    # We did this before by getting the list of group ids, and
-                    # then passing that list to sqlite to get latest event for
-                    # each (type, state_key). However, that was terribly slow
-                    # without the right indices (which we can't add until
-                    # after we finish deduping state, which requires this func)
-                    args = [next_group]
-                    args.extend(where_args)
-
-                    txn.execute(
-                        "SELECT type, state_key, event_id FROM state_groups_state"
-                        " WHERE state_group = ? " + where_clause,
-                        args,
-                    )
-                    results[group].update(
-                        ((typ, state_key), event_id)
-                        for typ, state_key, event_id in txn
-                        if (typ, state_key) not in results[group]
-                    )
-
-                    # If the number of entries in the (type,state_key)->event_id dict
-                    # matches the number of (type,state_keys) types we were searching
-                    # for, then we must have found them all, so no need to go walk
-                    # further down the tree... UNLESS our types filter contained
-                    # wildcards (i.e. Nones) in which case we have to do an exhaustive
-                    # search
-                    if (
-                        max_entries_returned is not None
-                        and len(results[group]) == max_entries_returned
-                    ):
-                        break
```

Comment:
With the current consolidation, we're losing this shortcut when running with SQLite 

And as mentioned in another comment, also conscious that some of these subtle hacks to consolidate might not be better than a big `if-else` distinction between Postgres and SQLite :grimacing:

Reply:
Closing as I'm not convinced this PR is better now that we've reached this point where things work. It was looking very promising before these latest changes necessitated by SQLite not supporting `SELECT DISTINCT ON` syntax.

It is less code especially considering comments added but wouldn't consider it easier to reason about and we lose the SQLite optimization mentioned above.